test-builds.sh

```diff
@@ -138,6 +138,13 @@ if [ -n "$cache_file" -a -e "$cache_file" -a "$remove_cache_file" = "true" ]; th
     rm $cache_file
 fi
 
+if test -e $top/SPONSORS && test -e $top/Makefile.in && test -e $top/libltdl/Makefile.in
```

Comment:
Four issues with this one line:
 1) Paths may contain whitespace. Especially given we support Windows builds. Please always quote the path parameters in shell scripts.

2) Since these tests are used a lot this is a prime place for test "-a" parameter performance gains vs shell "&&" to add up over time.

3) Using the test -f parameter to ensure these are actually files is more secure. It is only a good bootstrap if these are regular files autoconf created - not symlinks, directory names, pipes or other odd things in the FS.

4) [polish] the rest of the script uses "if condition; then" line wrapping. Please confirm to that for consistency.

```suggestion
if test -f "$top/SPONSORS" -a -f "$top/Makefile.in" -a -f "$top/libltdl/Makefile.in"; then
```

PS: I know test -e and lack of filepath quoting exists elsewhere in these scripts. Those are bugs that need fixing, but I do not insist on happening here.

Reply:
Bootstrapping produces 100+ files. If possible, please either clarify the principle behind selecting these three seemingly random files (via a source code comment) or use a different principle that would not require documentation.

I suggest using a single, and arguably the most obvious bootstrapping artifact for this "bootstrapping has happened" test, reducing the probability that it will become stale as the code changes:

```suggestion
if test -e "$top/configure"; then
```

As for `-e` vs. `-f`, unfortunately, I do not understand what security concerns Amos has -- no security concern that I can think of is both relevant to `test-builds.sh` _and_ can be addressed by using the more restrictive `-f`. FWIW, I prefer the more general `-e` check (e.g., it allows for occasionally useful Squid building hacks).

Reply:
The security issue is related to that promise we have for our scripts - that they only do what the repository sources say they do. For example; if the auto-generated files were allowed to be delivered to a pipe with some third-party logic controlling what automake receives when it reads that pipe it may seriously compromise our resulting binaries without any detectable source code hijack.

It is more serious problem for the production binary builds than for testing. But not something we should be letting happen either way.

Reply:
Thanks for commenting!
Re: spaces in paths
You are right, will fix

Re: -a vs &&
That bit of code gets executed exactly once per test build, I don't see it as performance-critical, therefore me going for the most portable option

Re: -e vs -f
I agree the most restrictive option can make sense. Hacks can create regular files there. The only other sensible options would be named pipes or fifos (not dirs or devices), and I see nothing there that can't be done by creating a file.

Re: polish
OK

Re: checking for configure instead of other artifacts
Sure.
I'd check two files, one in the topdir and one in topdir/libltdl as these are independently bootstrapped

Reply:
After discussing this PR with another developer, I can relay an even better alternative for you to consider -- test-builds.sh should call `bootstrap.sh` unconditionally [before each ./configure]:

1. Our test-builds.sh is supposed to test a few from-scratch builds (rather than at most one from-scratch builds plus a few partial builds). There should be no build-related "state" preserved between its iterations.
1. Compared to running any of the build tests, bootstrapping is a minor expense. 
1. Doing bootstrapping unconditionally would simplify things, address all the change requests still being discussed, and also remove "how/did we bootstrap correctly?" concerns when triaging test-buids.sh failures.

----

If you are sure we have to make bootstrapping conditional, please read on.

> Kinkie: I'd check two files, one in the topdir and one in topdir/libltdl as these are independently bootstrapped

After looking through bootstrap.sh, I have to conclude that the "independently bootstrapped" criteria is similar to the "3 seemingly random files" criteria critiqued earlier because `bootstrap.sh` does not know that calling bootstrap() with _some_ parameters creates additional `configure` files in subdirectories. The new "independently bootstrapped" concept would almost make sense if `./bootstrap.sh` called some `$dir/bootstrap.sh` (with each call expected to produce `$dir/configure`), but that is not happening AFAICT.

Besides, `ltdl` is a minor implementation detail from the high-level "Was this Squid bootstrapped?" point of view. Tomorrow, that directory may be gone or another may be added. None of those events should affect the code you are committing.

The fact that bootstrap.sh creates `./configure` is the only thing the bootstrap.sh user should know (because they are going to run `./configure` after running boostrap.sh). All other bootstrapping artifacts are a secondary implementation detail.

Thus, for the inferior conditional bootstrapping alternative, I continue to recommend using a single correct artifact for detecting bootstrapping -- `./configure`.

I do not have the energy to insist on this change.


> Amos: The security issue is related to that promise we have for our scripts - that they only do what the repository sources say they do.

Hm... That is not a promise, it is a fact -- our scripts do what they do! We had discussed a similar-sounding promise recently, but it was different (and irrelevant here AFAICT). Here, bootstrap.sh promises to prepare sources for running ./configure, while test-builds.sh promises to test a few build scenarios.

> For example; if the auto-generated files were allowed to be delivered to a pipe with some third-party logic controlling what automake receives when it reads that pipe it may seriously compromise our resulting binaries without any detectable source code hijack.

Agreed. I do not see the relevance though. Surely, a malicious third-party script that substitutes ./configure with a pipe can also edit that ./configure file in place to defeat our `-f` check. In other words, in this strawman argument, we have already lost at the "there is a powerful malicious third-party player on our build box" stage.

> It is more serious problem for the production binary builds than for testing. 

IMO, the problem you are describing is completely unrelated to detecting prior bootstrapping in test-builds.sh.

> But not something we should be letting happen either way.

As you know, I think it is exactly what we should let happen (in this PR context) because it enables occasionally useful tricks and does not harm regular setups.

However, after checking that `test -f` works for soft links, I can ignore this part of the issue! If you insist on being excessively strict in this test, I am not going to object to `-f`.